### ISSUE TITLE
SEGV when running MVM_SPESH_NODELAY=1 MVM_SPESH_BLOCKING=1 ./perl6 t/spec/S32-list/sort.t

### ISSUE THREAD

**Issue by dogbert17** [MoarVM/MoarVM#1027]
A recently introduced bug was uncovered by nwc10++. Running 

    MVM_SPESH_NODELAY=1 MVM_SPESH_BLOCKING=1 ./perl6-gdb-m t/spec/S32-list/sort.t

often leads to a SEGV.

Using his bisectable wizardry AlexDaniel++ managed to track down the introduction of the SEGV to the merge of the _cur-candidates_ branch. 

Note: it might be worth noting that the SEGV disappears if MVM_JIT_DISABLE=1.

    dogbert@dogbert-VirtualBox ~/repos/rakudo $ MVM_SPESH_NODELAY=1 MVM_SPESH_BLOCKING=1 ./perl6-gdb-m t/spec/S32-list/sort.t
    ================================================================================================
    This is Rakudo Perl 6 running in the GNU debugger, which often allows the user to generate useful back-
    traces to debug or report issues in Rakudo, the MoarVM backend or the currently running code.
    
    This Rakudo version is 2018.12.37.g.45.a.945.b built on MoarVM version 2018.12,
    running on linuxmint (18.3.Sylvia) / linux (4.10.0.38.generic)
    
    Type `bt full` to generate a backtrace if applicable, type `q` to quit or `help` for help.
    ------------------------------------------------------------------------------------------------
    Reading symbols from /home/dogbert/repos/rakudo/install/bin/moar...done.
    Starting program: /home/dogbert/repos/rakudo/install/bin/moar --execname=./perl6-gdb-m --libpath=. --libpath=blib --libpath=/home/dogbert/repos/rakudo/install/share/nqp/lib --libpath=/home/dogbert/repos/rakudo/install/share/nqp/lib --libpath=/home/dogbert/repos/rakudo/install/share/nqp/lib /home/dogbert/repos/rakudo/perl6.moarvm --nqp-lib=blib t/spec/S32-list/sort.t
    [Thread debugging using libthread_db enabled]
    Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
    [New Thread 0x7ffff6228700 (LWP 24619)]
    1..37
    ok 1 - array of numbers was sorted
    ok 2 - array of numbers was sorted (w/out parens)
    ok 3 - array of mixed numbers including Inf
    ok 4 - array of numbers was sorted (using invocant form)
    ok 5 - ... with explicit spaceship
    ok 6 - ... with explicit spaceship (using invocant form)
    ok 7 - ... reverse sort with explicit spaceship
    ok 8 - ... reverse sort with explicit spaceship (using invocant form)
    ok 9 - array of strings was sorted
    ok 10 - array of strings was sorted (w/out parens)
    ok 11 - array of strings was sorted (using invocant form)
    ok 12 - ... with explicit cmp
    ok 13 - ... with explicit cmp (using invocant form)
    ok 14 - ... sort keys by string value
    ok 15 - ... sort keys by string value (using invocant form)
    ok 16 - ... sort keys by numeric value
    ok 17 - ... sort keys by numeric value (using invocant form)
    ok 18 - can sort with automated Schwartzian Transform
    ok 19 - %hash.sort returns a List of Pairs
    ok 20 - sort with unary sub
    ok 21 - sort with arity 0 closure is stable
    ok 22 - sort with arity 1 closure is stable
    ok 23 - sort with arity 2 closure is stable
    ok 24 - method form of sort should work on numbers
    ok 25 - method form of sort should work on strings
    ok 26 - method form of sort should work on lists
    
    Thread 1 "moar" received signal SIGSEGV, Segmentation fault.
    0x00007ffff762f8a9 in MVM_string_check_arg (tc=0x604a50, s=0xffffffff00000001, operation=0x7ffff76fe34c "equal") at src/strings/ops.h:36
    36	    if (!s || !IS_CONCRETE(s))
    (gdb) bt
    #0  0x00007ffff762f8a9 in MVM_string_check_arg (tc=0x604a50, s=0xffffffff00000001, operation=0x7ffff76fe34c "equal") at src/strings/ops.h:36
    #1  0x00007ffff7633ce9 in MVM_string_equal (tc=0x604a50, a=0xffffffff00000001, b=0xf33290) at src/strings/ops.c:1426
    #2  0x00007ffff306a139 in ?? ()
    #3  0x00007ffff753749d in MVM_p6opaque_read_object (tc=0x911578, o=0x7fffffffc440, offset=6310480) at src/6model/reprs/P6opaque.h:115
    #4  0x00007ffff76a51c4 in MVM_jit_code_enter (tc=0x604a50, code=0x7fffec687c20, cu=0x690700) at src/jit/interface.c:24
    #5  0x00007ffff7530d0f in MVM_interp_run (tc=0x604a50, initial_invoke=0x7ffff76828fd <toplevel_initial_invoke>, invoke_data=0x69feb8) at src/core/interp.c:6146
    #6  0x00007ffff7682a62 in MVM_vm_run_file (instance=0x604010, filename=0x7fffffffe24d "/home/dogbert/repos/rakudo/perl6.moarvm") at src/moar.c:440
    #7  0x00000000004016e7 in main (argc=10, argv=0x7fffffffdd98) at src/main.c:301
    (gdb) info threads
      Id   Target Id         Frame 
    * 1    Thread 0x7ffff7fdb700 (LWP 24618) "moar" 0x00007ffff762f8a9 in MVM_string_check_arg (tc=0x604a50, s=0xffffffff00000001, operation=0x7ffff76fe34c "equal") at src/strings/ops.h:36
      2    Thread 0x7ffff6228700 (LWP 24619) "moar" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
    (gdb) p MVM_dump_backtrace(tc)
       at SETTING::src/core/Str.pm6:3060  (./CORE.setting.moarvm:infix:<ne>)
     from SETTING::src/core/Mu.pm6:549  (./CORE.setting.moarvm:Str)
     from SETTING::src/core/Mu.pm6:546  (./CORE.setting.moarvm:Str)
     from SETTING::src/core/Mu.pm6:566  (./CORE.setting.moarvm:Stringy)
     from SETTING::src/core/Mu.pm6:563  (./CORE.setting.moarvm:Stringy)
     from SETTING::src/core/Order.pm6:22  (./CORE.setting.moarvm:infix:<cmp>)
     from SETTING::src/core/Order.pm6:9  (./CORE.setting.moarvm:infix:<cmp>)
     from t/spec/S32-list/sort.t:189  (<ephemeral file>:)
     from SETTING::src/core/Rakudo/Sorting.pm6:100  (./CORE.setting.moarvm:MERGESORT-REIFIED-LIST-WITH)
     from SETTING::src/core/List.pm6:1395  (./CORE.setting.moarvm:sort)
     from t/spec/S32-list/sort.t:189  (<ephemeral file>:)
     from perl#sources/2D4A7CA10695CD2B374573413D261A660E282E2A (Test):535  (/home/dogbert/repos/rakudo/install/share/perl6/precomp/FC62179B46BB1BB88BADEBB0187A936F30518E99/2D/2D4A7CA10695CD2B374573413D261A660E282E2A:)
     from perl#sources/2D4A7CA10695CD2B374573413D261A660E282E2A (Test):534  (/home/dogbert/repos/rakudo/install/share/perl6/precomp/FC62179B46BB1BB88BADEBB0187A936F30518E99/2D/2D4A7CA10695CD2B374573413D261A660E282E2A:lives-ok)
     from perl#sources/2D4A7CA10695CD2B374573413D261A660E282E2A (Test):532  (/home/dogbert/repos/rakudo/install/share/perl6/precomp/FC62179B46BB1BB88BADEBB0187A936F30518E99/2D/2D4A7CA10695CD2B374573413D261A660E282E2A:lives-ok)
     from t/spec/S32-list/sort.t:189  (<ephemeral file>:)
     from t/spec/S32-list/sort.t:189  (<ephemeral file>:)
     from t/spec/S32-list/sort.t:187  (<ephemeral file>:<unit>)
     from t/spec/S32-list/sort.t:1  (<ephemeral file>:<unit-outer>)
     from gen/moar/stage2/NQPHLL.nqp:1685  (/home/dogbert/repos/rakudo/install/share/nqp/lib/NQPHLL.moarvm:eval)
     from gen/moar/stage2/NQPHLL.nqp:1928  (/home/dogbert/repos/rakudo/install/share/nqp/lib/NQPHLL.moarvm:evalfiles)
     from gen/moar/stage2/NQPHLL.nqp:1818  (/home/dogbert/repos/rakudo/install/share/nqp/lib/NQPHLL.moarvm:command_eval)
     from src/Perl6/Compiler.nqp:48  (blib/Perl6/Compiler.moarvm:command_eval)
     from gen/moar/stage2/NQPHLL.nqp:1777  (/home/dogbert/repos/rakudo/install/share/nqp/lib/NQPHLL.moarvm:command_line)
     from gen/moar/main.nqp:54  (/home/dogbert/repos/rakudo/perl6.moarvm:MAIN)
     from gen/moar/main.nqp:42  (/home/dogbert/repos/rakudo/perl6.moarvm:<mainline>)
     from <unknown>:1  (/home/dogbert/repos/rakudo/perl6.moarvm:<main>)
     from <unknown>:1  (/home/dogbert/repos/rakudo/perl6.moarvm:<entry>)
    $1 = void
    (gdb)

**Comment by AlexDaniel** [comment]
FWIW https://github.com/rakudo/rakudo/commit/d41bc6d is *bad* but both of its parents are *good*… so bisection doesn't give a helpful result.

**Comment by ugexe** [comment]
It is not immediately obvious to me how that would be caused by the cur-candidates branch. Some things to note:

---

Changing
```
use Test;
use lib $?FILE.IO.parent(2).add("packages/Test-Helpers");
use Test::Util;
```
to
```
use lib $?FILE.IO.parent(2).add("packages/Test-Helpers");
use Test::Util;
use Test;
```
makes it go away.

---

`%*ENV<RAKUDO_MODULE_DEBUG>=1;` makes it go away

**Comment by niner** [comment]
It's the JIT code for getcodename but I haven't found out why exactly yet.

**Comment by niner** [comment]
Fixed with commit caf7e4ff2cb49c8d85c8e6f32b2ab7646579c1a1

**Comment by AlexDaniel** [comment]
Thank you, @niner!
